### DISCUSSION TITLE
OCaml help, please: Trying to "if" and split a string.

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3145] [Q&A] [answered]
I’d like to implement a videostream with AzuraCast and got quite far, but can’t get my head around the OCaml language, so I ask a little help from the OCaml experts, please!

My problem is that when playling a playlist, I of course get metadata for artist and title in `m["artist"]` and `m["title"]` and my video will look like this:

![video m2t - VLC media player_003](https://github.com/savonet/liquidsoap/assets/3706922/2122d8f3-4149-4903-9991-56da534133bc)

So the title is in the first line of the lower-third, the artist in the second.

Whenever I relay a _stream_ (like from a syndication), the artist is blank and _both_ artist and title are shown in `m["title"]`, separated by ` - ` (blank, hyphen, blank):

![video m2t - VLC media player_002](https://github.com/savonet/liquidsoap/assets/3706922/d56fe51a-41e9-4445-82de-0a5207d3b055)

Ideally, it _should_ look like in the first picture, i.e., have "How Big, How Blue, How Beautiful…" in the _first line_, and "Florence + the Machine" in the _second line_.

I would need help in constructing an "if artist is blank" "then" "split title by the _first_ ' - ' (but leave further ' - ' alone) and put the first list element into `artist` and the second into `title`". For the time being, I currently simply concatenate these two to output a two-line string using `video.add_text()`.

My current code looks like:
```
# Add Now Playing metadata
nowplaying = ref("")
def update_nowplaying(m)
    nowplaying.set(m["title"] ^ "\n" ^ m["artist"])
end
radio.on_metadata(update_nowplaying)

…

s = video.add_text(font=font_file, size=40, x=50, y=580, color=0xebebeb, {nowplaying()}, s)
```

Help much appreciated! I tried and played around for quite a while but can’t get the hang of OCaml…

---

The second image shows another problem: When the metadata gets too long, it looks ugly, because I don’t know how to limit the generated text’s length (in pixels). As far as I know, there is no "max_width" in `video.add_text`. Since the text starts at an x offset of 50, I’d like to "cut" it at the right, also offset by 50 pixels; it should have a maximum length of 1280-50-50 = 1180 pixels. We must use pixels here, since we wouldn’t know how a certain font renders and thus can’t simply cut by number of characters.

Does anyone have a good idea how to achieve this?

**Comment by smimram** [comment, marked as answer]
I will add soon `string.split.first`, see #3146. With this you should be able to do something like
```
s = "abc - def"
if string.contains(substring=" - ", s) then
  let (artist, title) = string.split.first(separator=" - ", s)
  print("artist: #{artist}")
  print("title: #{title}")
end
```

**Comment by Moonbase59** [comment]
Sounds wonderful, thanks! Getting nearer to beautiful webradio video streams every day… :-)

Any idea how I could solve the text max. length (pixels) problem?

**Comment by Moonbase59** [comment]
@smimram Will the (currently failing) build automatically propagate to https://github.com/savonet/liquidsoap-release-assets/releases/tag/rolling-release-v2.2.x? (I would need the Ubuntu Jammy `.deb` to test locally and within the station.)

I have tried to adapt my code already, hopefully this will work:
```
# Add Now Playing metadata
nowplaying = ref("")
def update_nowplaying(m)
    title = m["title"]
    artist = m["artist"]
    if artist = "" then
        if string.contains(substring=" - ", title) then
            let (artist, title) = string.split.first(separator=" - ", title)
        end
    end
    #nowplaying.set("#{title}\n#{artist}")
    nowplaying.set(title ^ "\n" ^ artist)
end
a.on_metadata(update_nowplaying)
```

**Reply by smimram** [reply]
In order to test quickly you can just copy the code of `string.split.first` at the beginning of your file!

**Comment by Moonbase59** [comment]
Thanks, but the `.deb` was ready now. Only the above doesn’t work, I simply don’t know what is what (in spite of reading "The Book" and googling around for OCaml stuff). Keep getting a Parse error on the `let (artist, title)` and don’t really know how variables function… Must I use `=` or `:=` or `ref()` or `x.set()` for assignment, must I use `title` or `{title}` or `#{title}` to reference a variable?… Desperately playing around… Phew.

**Comment by Moonbase59** [comment]
Phew. Spent over two _hours_ just to find how to check for an empty string. And realized you’re right, Liquidsoap _isn’t OCaml_ (many OCaml examples simply won’t work). But thanks to a lot of learning, googling, reading "The Book" and the dev docs, and a zillion experiments, I finally managed. Hooray!

For strings being too long to display, I currently truncate the string. Not exact, but works for now.

Here’s what I currently use:
```
# Maximum length of text in characters (shorten will add a "…") -- depends on font used
font_maxlen = 53

…

# reduce volume target from -18 LUFS to EBU recommended -23 LUFS
a = amplify(0.562, a)

# Shorten string so it fits on screen (depends on font, must be measured)
def shorten(s, l=font_maxlen)
  if string.length(s) > l then
    # Using 3 periods instead of real ellipsis, because ellipsis isn’t fully displayed
    string.sub(s, start=0, length=l) ^ "..."
  else
    s
  end
end

# Add Now Playing metadata
# If we get an empty artist, assume a stream
# and try to split the title into artist and title.
nowplaying = ref("")
def update_nowplaying(m)
    nowplaying.set(shorten(m["title"]) ^ "\n" ^ shorten(m["artist"]))
    if string.length(m["artist"]) <= 0 then
      if string.contains(substring=" - ", m["title"]) then
          let (a, t) = string.split.first(separator=" - ", m["title"])
          nowplaying.set(shorten(t) ^ "\n" ^ shorten(a))
      end
    end
end

a.on_metadata(update_nowplaying)
```

May not be elegant, but works for now. Thanks again for implementing `string.split.first()`, this really helped.

![Florence + the Machine - How Big, How Blue, How Beautiful - mpv_002](https://github.com/savonet/liquidsoap/assets/3706922/739d41eb-9f05-49aa-bf87-56abf933fbe1)

Btw, the [`
liquidsoap-dcf405f_2.2.0-ubuntu-jammy-1_amd64.deb`](https://github.com/savonet/liquidsoap-release-assets/releases/download/rolling-release-v2.2.x/liquidsoap-dcf405f_2.2.0-ubuntu-jammy-1_amd64.deb) complained with some error when installing, seems to have something to do with the docs. Sorry, forgot the exact error message and can’t right now re-install because Liquidsoap currently running…
